File: omok/mcts.py
"""Batched PUCT Monte-Carlo tree search.

The trick that makes this usable in Python: many games are searched at once and
every game contributes one leaf per iteration, so each neural-network call sees
a batch of `len(trees)` positions instead of one.  Nothing here is
multi-threaded -- the parallelism is in the batch.
"""

from __future__ import annotations

import math
from typing import Callable, Sequence

import numpy as np

from .board import Board
from .config import MCTSConfig
from .encode import encode_batch


class Node:
    # `legal`/`bias`/`nf`/`unvisited` and the two scratch buffers exist purely to
    # keep `Tree.select_action` cheap: it runs tens of thousands of times per
    # search and every avoided array allocation shows up in the profile.
    __slots__ = ("P", "N", "W", "children", "expanded", "terminal", "visits", "value",
                 "legal", "bias", "nf", "unvisited", "q_buf", "u_buf",
                 "w_sum", "p_explored")

    def __init__(self) -> None:
        self.P: np.ndarray | None = None
        self.N: np.ndarray | None = None
        self.W: np.ndarray | None = None
        self.children: dict[int, "Node"] = {}
        self.expanded = False
        self.terminal = False
        self.visits = 0
        self.value = 0.0
        self.legal: np.ndarray | None = None
        self.bias: np.ndarray | None = None
        self.nf: np.ndarray | None = None
        self.unvisited: np.ndarray | None = None
        self.q_buf: np.ndarray | None = None
        self.u_buf: np.ndarray | None = None
        self.w_sum = 0.0
        self.p_explored = 0.0

    def expand(self, priors: np.ndarray, legal: np.ndarray) -> None:
        action_size = len(priors)
        masked = np.where(legal, priors, 0.0).astype(np.float32)
        total = float(masked.sum())
        if total <= 1e-8:  # network gave all its mass to illegal moves
            masked = legal.astype(np.float32)
            total = float(masked.sum())
        self.P = masked / max(total, 1e-8)
        self.N = np.zeros(action_size, dtype=np.int32)
        self.W = np.zeros(action_size, dtype=np.float32)
        # A node's position never changes, so its legal moves never change:
        # cache the mask instead of recomputing it on every descent, and keep
        # it as an additive 0 / -inf bias so scoring stays one vectorised add.
        self.legal = legal
        self.bias = np.where(legal, 0.0, -np.inf)
        self.nf = np.zeros(action_size, dtype=np.float64)
        self.unvisited = np.ones(action_size, dtype=np.float64)
        self.q_buf = np.empty(action_size, dtype=np.float64)
        self.u_buf = np.empty(action_size, dtype=np.float64)
        self.w_sum = 0.0
        self.p_explored = 0.0
        self.expanded = True

    def q_values(self) -> np.ndarray:
        assert self.N is not None and self.W is not None
        with np.errstate(invalid="ignore", divide="ignore"):
            return np.where(self.N > 0, self.W / np.maximum(self.N, 1), 0.0)


class Tree:
    """A search tree rooted at a position, reused across the moves of a game."""

    def __init__(self, board: Board, cfg: MCTSConfig) -> None:
        self.cfg = cfg
        self.board = board
        self.root = Node()
        self.root_noise_applied = False
        self._pending_path: list[tuple[Node, int]] | None = None
        self._pending_board: Board | None = None
        self._pending_node: Node | None = None

    # -- tree walking ------------------------------------------------------
    def select_action(self, node: Node) -> int:
        """PUCT: argmax over Q + U, with unvisited children held at the FPU.

        Identical arithmetic to the obvious formulation, but `w_sum` and
        `p_explored` are maintained by `backup` rather than re-reduced here,
        and the intermediates land in the node's own scratch buffers.
        """
        assert node.nf is not None and node.W is not None and node.P is not None
        visits = node.visits
        total = visits if visits > 0 else 1
        parent_q = node.w_sum / visits if visits > 0 else 0.0
        fpu = parent_q - self.cfg.fpu_reduction * math.sqrt(max(node.p_explored, 0.0))

        nf, q, u = node.nf, node.q_buf, node.u_buf
        np.maximum(nf, 1.0, out=q)
        np.divide(node.W, q, out=q)          # Q, and exactly 0 where unvisited
        np.add(nf, 1.0, out=u)
        np.divide(node.P, u, out=u)
        u *= self.cfg.c_puct * math.sqrt(total)
        q += u
        if fpu != 0.0:                        # unvisited children sit at the FPU
            q += node.unvisited * fpu
        q += node.bias                        # -inf on illegal moves
        return int(np.argmax(q))

    def descend(self) -> tuple[Node, Board, list[tuple[Node, int]]]:
        node = self.root
        board = self.board.copy()
        path: list[tuple[Node, int]] = []
        while node.expanded and not node.terminal:
            action = self.select_action(node)
            path.append((node, action))
            board.play(action)
            child = node.children.get(action)
            if child is None:
                child = Node()
                node.children[action] = child
            node = child
            if board.over:
                node.terminal = True
                node.value = board.result_for(board.to_move)
                break
        return node, board, path

    @staticmethod
    def backup(path: Sequence[tuple[Node, int]], value: float) -> None:
        sign = -1.0  # the leaf value is from the leaf mover's point of view
        for node, action in reversed(path):
            assert node.N is not None and node.W is not None
            if node.N[action] == 0:           # first visit: this child is now explored
                node.unvisited[action] = 0.0
                node.p_explored += float(node.P[action])
            node.N[action] += 1
            node.nf[action] += 1.0
            delta = sign * value
            node.W[action] += delta
            node.w_sum += delta
            node.visits += 1
            sign = -sign

    # -- one search iteration, split around the network call ---------------
    def begin_iteration(self) -> Board | None:
        """Descend to a leaf.  Returns the position needing evaluation, if any."""
        if self.board.over:
            return None
        node, board, path = self.descend()
        if node.terminal:
            self.backup(path, node.value)
            return None
        self._pending_node, self._pending_board, self._pending_path = node, board, path
        return board

    def finish_iteration(self, priors: np.ndarray, value: float,
                         rng: np.random.Generator | None = None) -> None:
        node, board, path = self._pending_node, self._pending_board, self._pending_path
        self._pending_node = self._pending_board = self._pending_path = None
        if node is None or board is None or path is None:
            return
        radius = self.cfg.prior_local_radius
        if radius > 0:
            local = np.where(board.neighbourhood(radius), priors, 0.0)
            if float(local.sum()) > 1e-6:
                priors = local
        node.expand(priors, board.legal_mask())
        node.value = float(value)
        if node is self.root and rng is not None:
            self.apply_root_noise(rng)
        self.backup(path, float(value))

    # -- root handling -----------------------------------------------------
    def apply_root_noise(self, rng: np.random.Generator) -> None:
        cfg = self.cfg
        if cfg.dirichlet_weight <= 0.0 or self.root.P is None or self.root_noise_applied:
            return
        legal = self.board.legal_mask()
        if cfg.prior_local_radius > 0:
            near = legal & self.board.neighbourhood(cfg.prior_local_radius)
            if near.any():
                legal = near
        indices = np.nonzero(legal)[0]
        if len(indices) == 0:
            return
        noise = rng.dirichlet([cfg.dirichlet_alpha] * len(indices)).astype(np.float32)
        priors = self.root.P.copy()
        priors[indices] = (1.0 - cfg.dirichlet_weight) * priors[indices] + cfg.dirichlet_weight * noise
        total = float(priors.sum())
        self.root.P = priors / max(total, 1e-8)
        self.root_noise_applied = True

    def visit_distribution(self) -> np.ndarray:
        if self.root.N is None:
            legal = self.board.legal_mask().astype(np.float32)
            return legal / max(legal.sum(), 1.0)
        counts = self.root.N.astype(np.float32)
        total = counts.sum()
        if total <= 0:
            legal = self.board.legal_mask().astype(np.float32)
            return legal / max(legal.sum(), 1.0)
        return counts / total

    def root_value(self) -> float:
        if self.root.N is None or self.root.visits == 0:
            return float(self.root.value)
        assert self.root.W is not None
        return float(self.root.W.sum() / max(self.root.visits, 1))

    def pick_move(self, rng: np.random.Generator, temperature: float) -> int:
        probs = self.visit_distribution()
        if temperature <= 1e-3:
            return int(np.argmax(probs))
        scaled = np.power(probs, 1.0 / temperature)
        total = scaled.sum()
        if total <= 0:
            return int(np.argmax(probs))
        return int(rng.choice(len(scaled), p=scaled / total))

    def advance(self, move: int) -> None:
        """Play ``move`` and keep the corresponding sub-tree."""
        child = self.root.children.get(move) if self.cfg.reuse_tree else None
        self.board = self.board.copy()
        self.board.play(move)
        self.root = child if child is not None else Node()
        self.root_noise_applied = False


Evaluator = Callable[[Sequence[Board]], tuple[np.ndarray, np.ndarray]]


class NetworkEvaluator:
    """Encodes positions and runs them through a backend, with a small cache."""

    def __init__(self, backend, batch_size: int = 256) -> None:
        self.backend = backend
        self.batch_size = batch_size
        self.calls = 0
        self.positions = 0

    def __call__(self, boards: Sequence[Board]) -> tuple[np.ndarray, np.ndarray]:
        if not boards:
            action_size = 0
            return np.zeros((0, action_size), np.float32), np.zeros((0,), np.float32)
        policies: list[np.ndarray] = []
        values: list[np.ndarray] = []
        for start in range(0, len(boards), self.batch_size):
            chunk = boards[start:start + self.batch_size]
            planes = encode_batch(chunk)
            policy, value = self.backend.predict(planes)
            policies.append(policy)
            values.append(value)
            self.calls += 1
            self.positions += len(chunk)
        return np.concatenate(policies), np.concatenate(values)


def run_search(trees: Sequence[Tree], evaluator: Evaluator, simulations: int,
               rng: np.random.Generator) -> None:
    """Run ``simulations`` iterations over all trees, batching the network calls."""
    active = [t for t in trees if not t.board.over]
    if not active:
        return
    # A reused sub-tree arrives with its root already expanded, so the noise
    # normally added on expansion must be injected here instead -- otherwise
    # self-play explores only on the first move of each game.
    for tree in active:
        if tree.root.expanded and not tree.root_noise_applied:
            tree.apply_root_noise(rng)
    for _ in range(simulations):
        pending: list[Tree] = []
        boards: list[Board] = []
        for tree in active:
            board = tree.begin_iteration()
            if board is not None:
                pending.append(tree)
                boards.append(board)
        if not pending:
            continue
        policies, values = evaluator(boards)
        for i, tree in enumerate(pending):
            tree.finish_iteration(policies[i], float(values[i]), rng)


def search_position(board: Board, evaluator: Evaluator, cfg: MCTSConfig,
                    simulations: int, rng: np.random.Generator,
                    add_noise: bool = False) -> Tree:
    """Convenience wrapper for searching a single position (play / analysis)."""
    tree = Tree(board.copy(), cfg)
    if not add_noise:
        tree.root_noise_applied = True  # suppress exploration noise
    run_search([tree], evaluator, simulations, rng)
    return tree

File: rust/crates/omok-core/src/mcts.rs
//! Batched PUCT search -- the Rust side of `omok/mcts.py`.
//!
//! The trainer batches its network calls across *games*: many self-play games
//! search at once, each contributing one leaf per iteration.  A game against a
//! human has only one position to think about, so this port batches across
//! *leaves* instead: it descends `batch` times before calling the network,
//! holding a virtual loss on each path so the descents pick different leaves.
//! Same arithmetic, same priors, same one-ply safety net -- but MLX sees
//! batches of 16 rather than a stream of single positions, which is the
//! difference between a search that feels instant and one that does not.

use crate::board::{Board, EMPTY};
use crate::net::Evaluator;
use crate::rng::Rng;

#[derive(Clone, Copy, Debug)]
pub struct MctsConfig {
    pub simulations: usize,
    pub c_puct: f32,
    pub fpu_reduction: f32,
    /// Zero the prior of moves farther than this from every stone (0 = off).
    pub prior_local_radius: usize,
    /// 0 = always play the most-visited move; higher samples more freely.
    pub temperature: f32,
    /// Leaves collected per network call.
    pub batch: usize,
}

impl Default for MctsConfig {
    fn default() -> Self {
        // The values the shipped checkpoint was trained with (runs/blitz).
        MctsConfig {
            simulations: 160,
            c_puct: 1.6,
            fpu_reduction: 0.25,
            prior_local_radius: 2,
            temperature: 0.0,
            batch: 16,
        }
    }
}

#[derive(Clone, Debug, Default)]
pub struct SearchResult {
    pub move_index: usize,
    /// Root value from the side-to-move's point of view, in [-1, 1].
    pub value: f32,
    pub simulations: usize,
    pub seconds: f64,
    /// Up to eight (move, visit share) pairs, best first.
    pub top: Vec<(usize, f32)>,
    /// True when a one-ply win or block overrode the search's choice.
    pub forced: bool,
}

#[derive(Default)]
struct Node {
    p: Vec<f32>,
    n: Vec<i32>,
    w: Vec<f32>,
    vloss: Vec<i32>,
    /// 0.0 on legal moves, -inf on illegal ones, so scoring stays one add.
    bias: Vec<f32>,
    /// 1.0 while a child has never been visited, 0.0 after.
    unvisited: Vec<f32>,
    /// action -> arena index, or `usize::MAX` for "not created yet".
    children: Vec<usize>,
    expanded: bool,
    terminal: bool,
    visits: i32,
    vloss_total: i32,
    value: f32,
    w_sum: f64,
    p_explored: f64,
}

const NO_CHILD: usize = usize::MAX;

impl Node {
    fn expand(&mut self, priors: &[f32], legal: &[bool]) {
        let action_size = priors.len();
        let mut masked: Vec<f32> = (0..action_size)
            .map(|i| if legal[i] { priors[i].max(0.0) } else { 0.0 })
            .collect();
        let mut total: f32 = masked.iter().sum();
        if total <= 1e-8 {
            // The network gave all its mass to illegal moves: fall back to
            // uniform over what is actually playable.
            masked = legal.iter().map(|&l| l as u8 as f32).collect();
            total = masked.iter().sum();
        }
        let scale = 1.0 / total.max(1e-8);
        self.p = masked.into_iter().map(|v| v * scale).collect();
        self.n = vec![0; action_size];
        self.w = vec![0.0; action_size];
        self.vloss = vec![0; action_size];
        self.bias = legal
            .iter()
            .map(|&l| if l { 0.0 } else { f32::NEG_INFINITY })
            .collect();
        self.unvisited = vec![1.0; action_size];
        self.children = vec![NO_CHILD; action_size];
        self.w_sum = 0.0;
        self.p_explored = 0.0;
        self.expanded = true;
    }
}

pub struct Search<'a> {
    cfg: MctsConfig,
    nodes: Vec<Node>,
    board: Board,
    rng: &'a mut Rng,
}

impl<'a> Search<'a> {
    pub fn new(board: Board, cfg: MctsConfig, rng: &'a mut Rng) -> Self {
        Search { cfg, nodes: vec![Node::default()], board, rng }
    }

    fn root(&self) -> usize {
        0
    }

    /// PUCT: argmax over Q + U, unvisited children held at the first-play
    /// urgency, illegal moves at -inf.  Virtual losses count as visits that
    /// lost, which is what keeps a batch of descents from piling onto one leaf.
    fn select_action(&self, idx: usize) -> usize {
        let node = &self.nodes[idx];
        let visits = node.visits + node.vloss_total;
        let total = if visits > 0 { visits } else { 1 } as f32;
        let sqrt_total = total.sqrt();
        let parent_q = if node.visits > 0 {
            (node.w_sum / node.visits as f64) as f32
        } else {
            0.0
        };
        let fpu = parent_q - self.cfg.fpu_reduction * (node.p_explored.max(0.0) as f32).sqrt();

        let mut best = 0usize;
        let mut best_score = f32::NEG_INFINITY;
        for a in 0..node.p.len() {
            if node.bias[a] == f32::NEG_INFINITY {
                continue;
            }
            let nf = (node.n[a] + node.vloss[a]) as f32;
            let w = node.w[a] - node.vloss[a] as f32;
            let q = w / nf.max(1.0);
            let u = self.cfg.c_puct * node.p[a] / (nf + 1.0) * sqrt_total;
            let mut score = q + u;
            if fpu != 0.0 {
                score += node.unvisited[a] * fpu;
            }
            if score > best_score {
                best_score = score;
                best = a;
            }
        }
        best
    }

    /// Walk to a leaf, creating nodes as needed.  Returns the leaf and the
    /// position it stands for, plus the (node, action) path taken.
    fn descend(&mut self) -> (usize, Board, Vec<(usize, usize)>) {
        let mut idx = self.root();
        let mut board = self.board.clone();
        let mut path = Vec::new();
        while self.nodes[idx].expanded && !self.nodes[idx].terminal {
            let action = self.select_action(idx);
            path.push((idx, action));
            board.play(action);
            let mut child = self.nodes[idx].children[action];
            if child == NO_CHILD {
                child = self.nodes.len();
                self.nodes.push(Node::default());
                self.nodes[idx].children[action] = child;
            }
            idx = child;
            if board.over {
                self.nodes[idx].terminal = true;
                self.nodes[idx].value = board.result_for(board.to_move);
                break;
            }
        }
        (idx, board, path)
    }

    fn add_virtual_loss(&mut self, path: &[(usize, usize)]) {
        for &(idx, action) in path {
            let node = &mut self.nodes[idx];
            node.vloss[action] += 1;
            node.vloss_total += 1;
        }
    }

    fn remove_virtual_loss(&mut self, path: &[(usize, usize)]) {
        for &(idx, action) in path {
            let node = &mut self.nodes[idx];
            node.vloss[action] -= 1;
            node.vloss_total -= 1;
        }
    }

    /// The leaf value is from the leaf mover's point of view, so it flips sign
    /// on every step back towards the root.
    fn backup(&mut self, path: &[(usize, usize)], value: f32) {
        let mut sign = -1.0f32;
        for &(idx, action) in path.iter().rev() {
            let node = &mut self.nodes[idx];
            if node.n[action] == 0 {
                node.unvisited[action] = 0.0;
                node.p_explored += node.p[action] as f64;
            }
            node.n[action] += 1;
            let delta = sign * value;
            node.w[action] += delta;
            node.w_sum += delta as f64;
            node.visits += 1;
            sign = -sign;
        }
    }

    /// Restrict the priors to the neighbourhood of the stones.  In gomoku
    /// every tactically relevant move is next to the action, so this points the
    /// whole simulation budget at the part of the board that matters.
    fn localise(&self, board: &Board, priors: &mut [f32]) {
        if self.cfg.prior_local_radius == 0 {
            return;
        }
        let near = board.neighbourhood(self.cfg.prior_local_radius);
        let kept: f32 = (0..priors.len()).map(|i| if near[i] { priors[i] } else { 0.0 }).sum();
        if kept > 1e-6 {
            for i in 0..priors.len() {
                if !near[i] {
                    priors[i] = 0.0;
                }
            }
        }
    }

    /// `stop` is polled once per network call, so a search can be abandoned
    /// the moment the player undoes a move or starts a new game.
    pub fn run(
        &mut self,
        evaluator: &mut dyn Evaluator,
        stop: &mut dyn FnMut() -> bool,
    ) -> Result<usize, String> {
        if self.board.over {
            return Ok(0);
        }
        let mut done = 0usize;
        let target = self.cfg.simulations.max(1);
        let batch = self.cfg.batch.max(1);

        while done < target {
            if stop() {
                break;
            }
            let want = batch.min(target - done);
            let mut pending: Vec<(usize, Board, Vec<(usize, usize)>)> = Vec::with_capacity(want);
            let mut claimed: Vec<usize> = Vec::with_capacity(want);
            let before = done;

            while pending.len() < want {
                let (leaf, board, path) = self.descend();
                if self.nodes[leaf].terminal {
                    let value = self.nodes[leaf].value;
                    self.backup(&path, value);
                    done += 1;
                    if done >= target {
                        break;
                    }
                    continue;
                }
                if claimed.contains(&leaf) {
                    // The virtual loss did not push this descent onto a
                    // different leaf, which happens when a node has only one
                    // move worth making.  Evaluate what has been gathered
                    // rather than expanding the same position twice.
                    break;
                }
                claimed.push(leaf);
                self.add_virtual_loss(&path);
                pending.push((leaf, board, path));
            }

            if pending.is_empty() {
                // Nothing to evaluate and nothing backed up: the tree cannot
                // grow any further from here, so stop instead of spinning.
                if done == before {
                    break;
                }
                continue;
            }

            let boards: Vec<&Board> = pending.iter().map(|(_, b, _)| b).collect();
            let planes = crate::encode::encode_batch(&boards);
            let prediction = evaluator.predict(&planes, boards.len())?;

            for (i, (leaf, board, path)) in pending.into_iter().enumerate() {
                self.remove_virtual_loss(&path);
                let mut priors = prediction.policy_row(i).to_vec();
                self.localise(&board, &mut priors);
                let value = prediction.value[i];
                let legal = board.legal_mask();
                self.nodes[leaf].expand(&priors, &legal);
                self.nodes[leaf].value = value;
                self.backup(&path, value);
                done += 1;
            }
        }
        Ok(done)
    }

    pub fn visit_distribution(&self) -> Vec<f32> {
        let root = &self.nodes[self.root()];
        if !root.expanded {
            let legal = self.board.legal_mask();
            let total = legal.iter().filter(|&&l| l).count().max(1) as f32;
            return legal.iter().map(|&l| l as u8 as f32 / total).collect();
        }
        let total: f32 = root.n.iter().map(|&n| n as f32).sum();
        if total <= 0.0 {
            let legal = self.board.legal_mask();
            let count = legal.iter().filter(|&&l| l).count().max(1) as f32;
            return legal.iter().map(|&l| l as u8 as f32 / count).collect();
        }
        root.n.iter().map(|&n| n as f32 / total).collect()
    }

    pub fn root_value(&self) -> f32 {
        let root = &self.nodes[self.root()];
        if !root.expanded || root.visits == 0 {
            return root.value;
        }
        (root.w_sum / root.visits.max(1) as f64) as f32
    }

    fn pick_move(&mut self, probs: &[f32]) -> usize {
        // The *first* maximum, as `np.argmax` returns and so as the Python
        // trainer picks.  It also keeps the chosen move in step with the top
        // list `finish` reports, which sorts stably: at a small simulation
        // budget a lot of moves tie on visits, and `max_by`'s last-maximum
        // would name a move that does not appear in the list at all.
        let argmax = |probs: &[f32]| {
            let mut best = 0usize;
            let mut best_prob = f32::NEG_INFINITY;
            for i in 0..probs.len() {
                if self.board.cells[i] == EMPTY && probs[i] > best_prob {
                    best_prob = probs[i];
                    best = i;
                }
            }
            best
        };
        if self.cfg.temperature <= 1e-3 {
            return argmax(probs);
        }
        let inv = 1.0 / self.cfg.temperature;
        let scaled: Vec<f32> = probs.iter().map(|&p| p.powf(inv)).collect();
        let total: f32 = scaled.iter().sum();
        if !(total > 0.0) {
            return argmax(probs);
        }
        let mut cursor = self.rng.next_f32() * total;
        for (i, &s) in scaled.iter().enumerate() {
            cursor -= s;
            if cursor <= 0.0 {
                return i;
            }
        }
        argmax(probs)
    }

    pub fn finish(mut self, simulations: usize, seconds: f64) -> SearchResult {
        let probs = self.visit_distribution();
        let mut order: Vec<usize> = (0..probs.len()).filter(|&i| probs[i] > 0.0).collect();
        order.sort_by(|&a, &b| probs[b].partial_cmp(&probs[a]).unwrap_or(std::cmp::Ordering::Equal));
        let top: Vec<(usize, f32)> = order.into_iter().take(8).map(|i| (i, probs[i])).collect();

        let value = self.root_value();
        // Never miss a one-ply win or fail to block one, whatever the search
        // says -- this is the difference between "beatable" and "broken".
        let forced = self.board.forced_move();
        let move_index = match forced {
            Some(m) => m,
            None => self.pick_move(&probs),
        };
        SearchResult {
            move_index,
            value,
            simulations,
            seconds,
            top,
            forced: forced.is_some(),
        }
    }
}

/// Search `board` and return the move to play.
pub fn search(
    board: &Board,
    evaluator: &mut dyn Evaluator,
    cfg: MctsConfig,
    rng: &mut Rng,
) -> Result<SearchResult, String> {
    search_interruptible(board, evaluator, cfg, rng, &mut || false)
}

/// As [`search`], but abandons the search as soon as `stop` returns true.
/// Returns `None` if it was stopped before a single simulation finished.
pub fn search_interruptible(
    board: &Board,
    evaluator: &mut dyn Evaluator,
    cfg: MctsConfig,
    rng: &mut Rng,
    stop: &mut dyn FnMut() -> bool,
) -> Result<SearchResult, String> {
    let started = std::time::Instant::now();
    let mut search = Search::new(board.clone(), cfg, rng);
    let simulations = search.run(evaluator, stop)?;
    Ok(search.finish(simulations, started.elapsed().as_secs_f64()))
}
